src/perl/lib/wtsi_clarity/epp/sm/fluidigm_analysis_file_creator.pm

```diff
@@ -0,0 +1,210 @@
+Readonly::Scalar our $SAMPLE_LIMSID            => q{./sample/@limsid};
+Readonly::Scalar our $CONTAINER_TYPE_URI       => q{/con:container/type/@uri};
+Readonly::Scalar our $CONTAINER_X_SIZE         => q{/ctp:container-type/x-dimension/size};
+Readonly::Scalar our $CONTAINER_Y_SIZE         => q{/ctp:container-type/y-dimension/size};
+Readonly::Scalar our $STA_PROCESS_NAME         => q{Fluidigm STA Plate Creation (SM)};
+
+has 'filename' => (
+  is       => 'ro',
+  isa      => 'Str',
+  required => 1,
+);
+
+override 'run' => sub {
+  my $self = shift;
+  super();
+
+  my $fluidigm_analysis_file = wtsi_clarity::genotyping::fluidigm_analysis->new(
+    sample_plate => $self->_sample_plate,
+    barcode      => $self->_barcode,
+    samples      => $self->_samples,
+  );
+
+  $fluidigm_analysis_file->saveas(q{./} . $self->filename);
+
+  return;
+};
+
+has '_sta_plate_creation_process' => (
+  is => 'ro',
+  isa => 'wtsi_clarity::clarity::process',
+  required => 0,
+  lazy_build => 1,
+);
+
+sub _build__sta_plate_creation_process {
+  my $self = shift;
+
+  my $processes = $self->process_doc->find_parent($STA_PROCESS_NAME, $self->process_url);
+
+  if (scalar @{$processes} == 0) {
+    croak "Can not find STA Plate";
+  }
+
+  my $process_xml = $self->fetch_and_parse($processes->[0]);
+
+  return wtsi_clarity::clarity::process->new(xml => $process_xml, parent => $self);
+}
+
+has ['_sample_plate', '_barcode'] => (
+  is         => 'ro',
+  isa        => 'Str',
+  init_arg   => undef,
+  lazy_build => 1,
+);
+
+sub _build__sample_plate {
+  my $self = shift;
+  return $self->_sta_plate_creation_process->plate_io_map_barcodes->[0]->{'source_plate'};
+}
+
+sub _build__barcode {
+  my $self = shift;
+  return $self->process_doc->plate_io_map_barcodes->[0]->{'source_plate'};
+}
+
+has '_samples' => (
+  is => 'ro',
+  isa => 'ArrayRef[HashRef]',
+  init_arg => undef,
+  lazy_build => 1,
+);
+
+sub _build__samples {
+  my $self = shift;
+  my @samples = ();
+
+  my $chr_num = ord('A') + $self->_container_size->{'y_dimension'} - 1;
+  my $big_row = chr $chr_num;
+
+  for (1..$self->_container_size->{'x_dimension'}) {
+    foreach my $column ('A'..$big_row) {
+
+      my $clarity_well = join q{:}, $column, $_;
+      my $fluidigm_well = $column . sprintf '%02d', $_;
+
+      my %sample = ();
+
+      my $artifact_list = $self->_sta_plate_creation_process
+                          ->input_artifacts
+                          ->findnodes(sprintf $ARTIFACT_BY_SAMPLE_VALUE, $clarity_well);
+
+      $sample{'well_location'} = $fluidigm_well;
+
+      if ($artifact_list->size == 0) {
+        $sample{'sample_name'} = '[ Empty ]';
+
+        if ($sample{'well_location'} eq 'H12') {
+          $sample{'sample_type'} = 'NTC';
+        }
+      } elsif ($artifact_list->size == 1) {
+        $sample{'sample_name'} = $artifact_list->pop->findvalue($SAMPLE_LIMSID);
+      }
+
```

Comment:
I've made the assumption that Clarity would never give us a plate with duplicate wells on it, but I'll add croak anyway